UPGRADE.md

```diff
@@ -44,6 +44,33 @@ before finalizing the upgrade process.
 1. ORY Oathkeeper now supports multiple mutators. Mutations are performed in the provided order and must all succeed in order for the HTTP request to be forwarded.
 2. The `mutator` property was renamed to `mutators` to reflect its true nature (see previous item).
 
+#### id_token mutator
+
+The id_token mutator now allows to specify custom claims via the `claims` field of the mutator's `config` field. The keys represent names of claims and the values are strings which will be parsed by the Go [text/template](https://golang.org/pkg/text/template/) package for value substitution, receiving the `AuthenticationSession` struct. You must now specify
+the audience of the ID token directly via the `aud` claim. If you use `id_token` mutators with defined audience, please update them as follows:
+
+deprecated config:
+```json
+{
+  "handler": "id_token",
+  "config": {
+    "aud": ["https://my-backend-service/some/endpoint"]
+  }
+}
+```
+
+new config:
+```json
+{
+  "handler": "id_token",
+  "config": {
+    "claims": {
+      "aud": ["https://my-backend-service/some/endpoint"]
```

Comment:
I think this makes it particularly hard to make use of these claims. I think the escape workaround ( https://github.com/ory/oathkeeper/pull/243/files#diff-5ab69f721ec10aff3758f597a3aa0de6R84 ) is also related to this. I had another idea, what do you think about using a raw json object in claims? So then you could do:

```json
{
  "aud": {{ .Whatever }}
}
```

or for a full example (**note: this needs sprig or "quote" will not work**):

```go
package main

import (
	"os"
	"text/template"
)

var card2 = map[string]string{"a": "whatever", "b": "bazbar"}

func check(err error) {
	if err != nil {
		panic(err)
	}
}

func main() {
	// Use {{define}} to prevent write multiple times.
	t := template.Must(template.New("t2").Parse(`{
	"aud": [{{ .whatever | quote }}, {{ .b | quote }}]
}`))
	check(t.Execute(os.Stdout, card2))
}
```

So basically `claims` would no longer be a `map[string]string` but instead just raw bytes. I think it makes sense to send the compiled template using `debug` log level if the compilation of the template fails.

This should then also help resolve and/or debug the e2e test failure.

What do you think?

Reply:
The above example would compile to:

```json
{
  "aud": ["whatever", "bazbar"]
}
```

Reply:
Using raw bytes IS the problem, I'm afraid. Parse func accepts a template like
```
`"{{- (index .Extra \"example.com/some-claims\") | join \",\" -}}"`
```
only if it has been decoded to string. Decoding claims to interface{} or even a structure like
```
type CredentialsIDTokenConfig struct {
	Claims json.RawMessage `json:"claims"`
}
```
preserves escape characters and causes Parse() to fail due to unexpected operands.

Reply:
see example: https://play.golang.org/p/m_9E7DlvHir

Reply:
Using raw bytes IS the problem, I'm afraid. Parse() refuses to digest something like
```
`{"custom-claim": "{{- (index .Extra \"example.com/some-claims\") | join \",\" -}}"}`
```
due to unexpected "\\" in operand. That's reasonable, since the sequence contains escape characters. If unmarshalled to a string, however, the
```
"{{- (index .Extra \"example.com/some-claims\") | join \",\" -}}"
```
expression can be parsed (escaped characters have been removed,) but bear in mind we allow arbitrary data structures, not just strings.

Reply:
Obviously, we could drop them backslashes from the config itself:
```
`{"claims":{"custom-claim": "{{- (index .Extra "example.com/some-claims") | join "," -}}", "aud": ["foo", "bar"]}}`
```
but they are required to unmarshal the object properly, so decoding simply won't work. 
We could even skip decoding by passing the entire "config" object (with no backslashes, as above) to Parse(). Template will be executed just fine, but this approach involves config validation (we allow only one key in the top-level map, i.e. "claims") and requires some further type assertions.

Reply:
This works fine though?

```go
package main

import (
	"os"
	"text/template"
)

type card struct {
	Extra map[string]interface{}
}

const jtemplate = `{
	"custom-claim": "{{- (index .Extra "example.com/some-claims") -}}"
}`

func main() {
	t := template.Must(template.New("t2").Parse(jtemplate))
	_ = t.Execute(os.Stdout, &card{Extra: map[string]interface{}{"example.com/some-claims": []string{"foo", "bar"}}})
}
```

Prints:

```
{
	"custom-claim": "[foo bar]"
}
```

Am I missing something?

Reply:
The example works fine, cause no decoding is involved. If you try to unmarshal a complete, unescaped "claims" object like this
```
`{"claims":{"custom-claim": "{{- (index .Extra \"example.com/some-claims\") | join \",\" -}}", "aud": ["foo", "bar"]}}`
```
the decoder simply goes nuts, because it expects a comma or another suitable conclusion to a key:value pair ("custom-claim": **"{{- (index .Extra \"**), but gets "example" instead.
See https://play.golang.org/p/jupXW2QTk8y.

The alternative involves parsing the config directly. In fact, it is going to work, but leaves us with a raw message (executed template) that still requires unmarshalling:
https://play.golang.org/p/v01iH3HLdig.

Is the second approach acceptable?

Reply:
The first example panics because jtemplate is the Go template - the json en/decoder can only work raw json, not go templates. Therefore it is necessary to compile the template first and then de/encode it to json! 

The second approach is more than acceptable, I think it's the only solution to this problem :)